File: frontend/src/api/factor.ts
/**
 * 因子研究 domain（2.3 G9 FactorLab）：
 *  - GET /factors —— 因子定义池（含 version/status/weight，构建器共用）
 *  - GET /factors/:name/stats —— 单因子检验统计（IC 时序/ICIR/衰减/分层/单调性），一次性给齐
 *  - GET /factors/stats/correlation —— 6 因子相关矩阵（区间均值）
 * 数据源：quant-engine 预计算 factor_stat/factor_corr 表（Go 仅读表 + 轻聚合，IC 数学单实现）。
 */
import { http } from './http'
import type { FactorCorrData, FactorsData, FactorStatsData } from './types'

export interface FactorStatsQuery {
  start?: string // YYYY-MM-DD，缺省近 2 年
  end?: string
  horizon?: number // {1,5,10,20,60}，缺省 5
}

export const factorApi = {
  /** 因子定义池（FactorLab 卡片 + 构建器共用） */
  getFactors: () => http.get<FactorsData>('/factors'),
  /** 单因子检验统计 */
  getFactorStats: (name: string, q?: FactorStatsQuery) =>
    http.get<FactorStatsData>(`/factors/${name}/stats`, { ...q }),
  /** 6 因子相关矩阵（缺省近 1 年） */
  getFactorCorr: (q?: FactorStatsQuery) =>
    http.get<FactorCorrData>('/factors/stats/correlation', { ...q }),
}

File: backend/internal/api/handler/factor.go
package handler

import (
	"errors"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"quant-system/backend/internal/service"
	"quant-system/backend/pkg/response"
)

// factorStatsDefaultYears 无 start/end 时默认回看区间（对齐前端 FactorLab 默认 2 年）
const factorStatsDefaultYears = 2

// parseFactorRange 解析 start/end/horizon（均可省略）
// start 默认 end-2 年、end 默认今天（服务端本地日）、horizon 默认 5。
func parseFactorRange(c *gin.Context) (start, end time.Time, horizon int, err error) {
	horizon = 5
	if h := c.Query("horizon"); h != "" {
		horizon, err = strconv.Atoi(h)
		if err != nil {
			return time.Time{}, time.Time{}, 0, errors.New("horizon 应为数字")
		}
	}
	end = time.Now().Truncate(24 * time.Hour)
	if s := c.Query("end"); s != "" {
		end, err = parseDate(s)
		if err != nil {
			return time.Time{}, time.Time{}, 0, errors.New("end 格式应为 YYYY-MM-DD")
		}
	}
	start = end.AddDate(0, 0, -factorStatsDefaultYears*365)
	if s := c.Query("start"); s != "" {
		start, err = parseDate(s)
		if err != nil {
			return time.Time{}, time.Time{}, 0, errors.New("start 格式应为 YYYY-MM-DD")
		}
	}
	return start, end, horizon, nil
}

// parseFactorDates 解析 start/end（correlation 无 horizon；缺省同 parseFactorRange）
func parseFactorDates(c *gin.Context) (start, end time.Time, err error) {
	end = time.Now().Truncate(24 * time.Hour)
	if s := c.Query("end"); s != "" {
		end, err = parseDate(s)
		if err != nil {
			return time.Time{}, time.Time{}, errors.New("end 格式应为 YYYY-MM-DD")
		}
	}
	start = end.AddDate(0, 0, -factorStatsDefaultYears*365)
	if s := c.Query("start"); s != "" {
		start, err = parseDate(s)
		if err != nil {
			return time.Time{}, time.Time{}, errors.New("start 格式应为 YYYY-MM-DD")
		}
	}
	return start, end, nil
}

// factorError 因子统计业务错误映射
func factorError(err error) (int, int, string) {
	switch {
	case errors.Is(err, service.ErrFactorNotFound):
		return http.StatusNotFound, response.CodeResourceMissing, err.Error()
	case errors.Is(err, service.ErrFactorHorizon),
		errors.Is(err, service.ErrFactorRange),
		errors.Is(err, service.ErrFactorSpan):
		return http.StatusBadRequest, response.CodeInvalidParam, err.Error()
	default:
		return http.StatusInternalServerError, response.CodeInternalError, "查询失败"
	}
}

// GetFactorStats 因子检验统计（GET /factors/:name/stats?start=&end=&horizon=）
// 契约《因子研究闭环》§6.1：ic_series/icir/ic_mean/ic_std/ic_decay/quantiles/monotonic。
func GetFactorStats(statsSvc *service.FactorStatsService) gin.HandlerFunc {
	return func(c *gin.Context) {
		start, end, horizon, err := parseFactorRange(c)
		if err != nil {
			response.Fail(c, http.StatusBadRequest, response.CodeInvalidParam, err.Error())
			return
		}
		out, err := statsSvc.Stats(c.Param("name"), start, end, horizon)
		if err != nil {
			status, code, msg := factorError(err)
			response.Fail(c, status, code, msg)
			return
		}
		response.OK(c, out)
	}
}

// GetFactorCorrelation 6 因子相关矩阵（GET /factors/stats/correlation?start=&end=）
// 契约 §6.1：{"factors": 6 因子规范序, "matrix": 6×6 区间均值（无共现格为 null）}。
func GetFactorCorrelation(statsSvc *service.FactorStatsService) gin.HandlerFunc {
	return func(c *gin.Context) {
		start, end, err := parseFactorDates(c)
		if err != nil {
			response.Fail(c, http.StatusBadRequest, response.CodeInvalidParam, err.Error())
			return
		}
		out, err := statsSvc.Correlation(start, end)
		if err != nil {
			status, code, msg := factorError(err)
			response.Fail(c, status, code, msg)
			return
		}
		response.OK(c, out)
	}
}
